### PR TITLE
Bug fix: hide pipeline section in add form when user has no access to list pipeline from openshift namespace

### PR DESCRIPTION
**Pull request by vikram-raj**
Fix bug - https://jira.coreos.com/browse/ODC-2373

Hide pipeline section in add form when the user has no access to list the pipeline from ```openshift``` namespace. 

The only user which has ```basic-role``` binding can not list the pipeline from the ```openshift``` namespace.

### CODE REVIEW

File: frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx [hunk 1]
```diff
@@ -15,7 +16,14 @@ type PipelineSectionProps = {
 const PipelineSection: React.FC<PipelineSectionProps> = ({ flags }) => {
   const { values } = useFormikContext<FormikValues>();
 
-  if (flags[FLAG_OPENSHIFT_PIPELINE]) {
+  const canListPiplines = useAccessReview({
+    group: PipelineModel.apiGroup,
+    resource: PipelineModel.plural,
+    namespace: 'openshift',
+    verb: 'list',
+  });
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx]: At the end of this, the user will need to be able to create a new pipeline instance as well. Shouldn't we also ensure they can create the pipeline?

File: frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx [hunk 2]
```diff
@@ -15,7 +16,14 @@ type PipelineSectionProps = {
 const PipelineSection: React.FC<PipelineSectionProps> = ({ flags }) => {
   const { values } = useFormikContext<FormikValues>();
 
-  if (flags[FLAG_OPENSHIFT_PIPELINE]) {
+  const canListPiplines = useAccessReview({
+    group: PipelineModel.apiGroup,
+    resource: PipelineModel.plural,
+    namespace: 'openshift',
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx]: I think we also need to stop doing this for our own sake. If the powers that be decide `openshift` is getting too crowded and it should be in something like `openshift-pipelines`, we need to be able to not start a scavanging hunt for the use of `'openshift'`. Please create a constant in `dev-console/src/const.ts` and use it here as well as update the reference [here](https://github.com/openshift/console/blob/master/frontend/packages/dev-console/src/components/import/pipeline/PipelineTemplate.tsx#L35).

Suggested variable name (feel free to use something else if you can think of a better name):
```
export const CLUSTER_PIPELINE_NS = 'openshift';
```

File: frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx (hunk 1 above)
Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx]: @christianvogt We have not create pipeline option in add form.

Comment (andrewballantyne) [reply frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx]: @vikram-raj There is a creation when you select `Add Pipeline` and [submit the form](https://github.com/openshift/console/blob/master/frontend/packages/dev-console/src/components/import/import-submit-utils.ts#L423).

Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx]: Ah ok. I will add RBAC check for create pipeline as well.

Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx]: @christianvogt Added RBAC checks for create pipeline access.

File: frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx (hunk 2 above)
Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx]: @andrewballantyne Added ```openshift``` in ```CLUSTER_PIPELINE_NS``` const.

File: frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx [hunk 3]
```diff
@@ -15,7 +17,21 @@ type PipelineSectionProps = {
...
+    namespace: CLUSTER_PIPELINE_NS,
+    verb: 'list',
+  });
+
+  const canCreatePiplines = useAccessReview({
+    group: PipelineModel.apiGroup,
+    resource: PipelineModel.plural,
+    namespace: getActiveNamespace(),
+    verb: 'create',
+  });
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx]: I think we also want to see if we can create `PipelineResources`, we [create them at the same time](https://github.com/openshift/console/blob/master/frontend/packages/dev-console/src/components/import//pipeline/pipeline-template-utils.ts#L45-L46). I doubt there _can_ be different permissions involved, but I think we should look to validate all the calls that are used so we don't have an issue in the future (or if you can control permissions that granular).

Recommend you refactor your checks into a "check for Pipeline creation permissions" method. That way it's isolated and this if-statement below doesn't become 40 boolean checks :)

Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx]: ```useAccessReview``` is a hook and we can not call it inside a method. we have to check it separately. I added a check for pipeline resources creation as well.

Comment (andrewballantyne) [reply frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx]: I thought Hooks could be called in any number of methods, it just had to always be invoked in the same order. Just start the method name with `use` to allow the linter to help out.

File: frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx [hunk 4]
```diff
@@ -15,7 +17,30 @@ type PipelineSectionProps = {
 const PipelineSection: React.FC<PipelineSectionProps> = ({ flags }) => {
   const { values } = useFormikContext<FormikValues>();
 
-  if (flags[FLAG_OPENSHIFT_PIPELINE]) {
+  const canListPiplines = useAccessReview({
```
Comment (debsmita1) [inline frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx]:
```suggestion
  const canListPipelines = useAccessReview({
```

File: frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx [hunk 5]
```diff
@@ -15,7 +17,30 @@ type PipelineSectionProps = {
...
+    group: PipelineModel.apiGroup,
+    resource: PipelineModel.plural,
+    namespace: CLUSTER_PIPELINE_NS,
+    verb: 'list',
+  });
+
+  const canCreatePiplines = useAccessReview({
```
Comment (debsmita1) [inline frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx]:
```suggestion
  const canCreatePipelines = useAccessReview({
```

File: frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx [hunk 6]
```diff
@@ -15,7 +17,30 @@ type PipelineSectionProps = {
...
+    group: PipelineModel.apiGroup,
+    resource: PipelineModel.plural,
+    namespace: getActiveNamespace(),
+    verb: 'create',
+  });
+
+  const canCreatePiplineResource = useAccessReview({
```
Comment (debsmita1) [inline frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx]:
```suggestion
  const canCreatePipelineResource = useAccessReview({
```

**Comment by debsmita1** [comment]
tested it locally, works fine

File: frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx (hunk 3 above)
Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx]: @andrewballantyne I added the method to handle all the checks.

File: frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx [hunk 7]
```diff
@@ -2,20 +2,49 @@ import * as React from 'react';
...
+  });
+
+  return canListPipelines && canCreatePipelines && canCreatePipelineResource;
+};
+
 const PipelineSection: React.FC<PipelineSectionProps> = ({ flags }) => {
   const { values } = useFormikContext<FormikValues>();
 
-  if (flags[FLAG_OPENSHIFT_PIPELINE]) {
+  const canCreate = useHasCreatePipelineAccess();
```
Comment (debsmita1) [inline frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx]: or may be: 
```suggestion
  const hasCreatePipelineAccess = usePipelineAccessReview();
```

Comment (andrewballantyne) [reply frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx]: Agreed, this is cleaner. Everything else looks good to me, just would like this change as well before providing approval.

Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/import/pipeline/PipelineSection.tsx]: @debsmita1 @andrewballantyne Done.

**Comment by andrewballantyne** [comment]
@vikram-raj you need to rebase. Once you do so I can add the labels back.

**Comment by christianvogt** [comment]
@rohitkrai03 according to Siamak, we don't need to backport the pipeline fix. Once their operator is fixed, that scenario won't occur anymore.

**Pull request merged**